hilti/runtime/include/types/optional.h

```diff
@@ -2,93 +2,383 @@
+        requires(! std::is_same_v<std::decay_t<U>, Optional> && std::is_constructible_v<T, U>)
+    {
+        std::optional<T>::emplace(std::forward<U>(v));
+        return *this;
+    }
+
+    /**
+     * Assigns from a value of different optional type containing a value of
+     * type U that's convertible to T.
+     */
+    template<typename U>
+    Optional& operator=(const Optional<U>& v)
+        requires(! std::is_same_v<U, T> && std::is_constructible_v<T, const U&>)
+    {
+        if ( v )
+            std::optional<T>::emplace(v.value());
+        else
+            reset();
+
+        return *this;
+    }
+
+    /**
+     * Assigns from a different optional type containing a value of type U
+     * that's convertible to T.
+     */
+    template<typename U>
+    Optional& operator=(Optional<U>&& v)
+        requires(! std::is_same_v<U, T> && std::is_constructible_v<T, U>)
+    {
+        if ( v )
+            std::optional<T>::emplace(std::move(v).value());
+        else
+            reset();
+
+        return *this;
+    }
+
+    template<typename U>
+    constexpr bool operator==(const Optional<U>& other) const {
+        if ( ! hasValue() && ! other.hasValue() )
+            return true;
+
+        if ( ! hasValue() || ! other.hasValue() )
+            return false;
+
+        return value() == other.value();
+    }
+
+    template<typename U>
+    constexpr bool operator!=(const Optional<U>& other) const {
+        return ! (*this == other);
+    }
+
+    template<typename U>
+    constexpr auto operator<(const Optional<U>& other) const {
+        if ( ! hasValue() && ! other.hasValue() )
+            return false;
+
+        if ( ! hasValue() )
+            return true;
+
+        if ( ! other.hasValue() )
+            return false;
+
+        return value() < other.value();
+    }
+
+    constexpr bool operator==(hilti::rt::Null) const noexcept { return ! hasValue(); }
+    constexpr bool operator!=(hilti::rt::Null) const noexcept { return hasValue(); }
+};
+
+namespace detail::adl {
+template<typename T>
+inline std::string to_string(const Optional<T>& x, adl::tag /*unused*/) {
+    return x ? hilti::rt::to_string(*x) : "(not set)";
 }
 
+} // namespace detail::adl
+
+namespace optional {
+
+/**
+ * Constructs an optional initialized to a given value. This is similar to
+ * ``std::make_optional``.
+ */
 template<class T>
-inline auto& tryValue(const std::optional<T>& t) {
-    if ( t.has_value() )
-        return t.value();
-    else
-        detail::throw_unset();
+constexpr Optional<std::decay_t<T>> make(T&& v) {
+    return Optional<std::decay_t<T>>(std::forward<T>(v));
+}
+
+/**
+ * Constructs an optional initialized to a value constructed from given
+ * arguments. This is similar to ``std::make_optional``.
+ */
+template<class T, class... Args>
+constexpr Optional<T> make(Args&&... args) {
+    return Optional<T>(std::in_place, std::forward<Args>(args)...);
 }
```

Comment:
In contrast to the `Tuple` functions these don't add anything over constructors, and deduction seems to work the same, e.g., this works:

```cpp
auto x = Optional(1); // Optional<int>
```

Can we omit these?